File: GoLang/utils/mathUtils.go
package utils

import (
	"math/rand"
	"time"
)

func ModPow(a, x, p int64) int64 {
    if p == 1 {
        return 0
    }

    result := int64(1)
    a = a % p
    
    for x > 0 {
        if x % 2 == 1 {
            result = (result * a) % p
        }
        x = x >> 1
        a = (a * a) % p
    }
    return result
}

func FindGcd(a int64, b int64) int64 {
	/* Находит НОД(a, b) */
	if b == 0 {
		return a
	}
	return FindGcd(b, a % b)
}

func ExtendedEuclid(a int64, b int64) (uint64, int64, int64) {
	/* Расширенный алгоритм Евклида */
	if b == 0 {
		return uint64(a), 1, 0
	}

	gcd, u1, v1 := ExtendedEuclid(b, a % b)
	u := v1
	v := u1 - int64(a / b) * v1

	return gcd, u, v
}

func EratosthenesSieve(limit int) []int64 {
	/* Строит решето Эратосфена */
	if limit < 2 {
		return []int64{}
	}

	isPrime := make([]bool, limit+1)
	for i := range isPrime {
		isPrime[i] = true
	}
	isPrime[0], isPrime[1] = false, false

	for p := 2; p*p <= limit; p++ {
		if isPrime[p] {
			for i := p * p; i <= limit; i += p {
				isPrime[i] = false
			}
		}
	}

	var primes []int64
	for i := 2; i <= limit; i++ {
		if isPrime[i] {
			primes = append(primes, int64(i))
		}
	}

	return primes
}

func Factorize(num int64) [][2]int64 {
	/* Разложение числа в канонический вид */
	primes := EratosthenesSieve(500)
	var result [][2]int64

	for _, prime := range primes {
		if prime > num {
			break
		}
		var exponent int64 = 0
		for num%prime == 0 && num > 0 {
			exponent++
			num /= prime
		}
		if exponent > 0 {
			result = append(result, [2]int64{prime, exponent})
		}
	}
	return result
}

func MillerTest(n int64, t int) bool {
	/* Вероятностный тест Миллера */
	if n == 2 || n == 3 || n == 5 || n == 7 {
		return true
	}
	if n < 10 || n%2 == 0 {
		return false
	}

	seed := time.Now().UnixNano()
	rng := rand.New(rand.NewSource(seed))

	nums := make(map[int64]bool)
	for len(nums) < t {
		candidate := rng.Int63n(n-2) + 2
		nums[candidate] = true
	}

	for a := range nums {
		if ModPow(a, n-1, n) != 1 {
			return false
		}
	}

	primeFactors := Factorize(n - 1)
	for _, factor := range primeFactors {
		q := factor[0]
		onlyOne := true
		for a := range nums {
			if ModPow(a, (n-1)/q, n) != 1 {
				onlyOne = false
				break
			}
		}
		if onlyOne {
			return false
		}
	}
	return true
}

func GenerateRandomPrime(bitLength uint) int64 {
    /* Генерация случайного числа длины bitLength */
    seed := time.Now().UnixNano()
    rng := rand.New(rand.NewSource(seed))

    for {
        var candidate int64 = rng.Int63()
        candidate |= 1
        candidate |= 1 << (bitLength - 1)

        if MillerTest(candidate, 8) {
            return candidate
        }
    }
}

File: C++/src/utils/mathUtils.cpp
#include <set>
#include <tuple>
#include <cmath>
#include <ctime>
#include <random>
#include <gmpxx.h>

#include "mathUtils.hpp"

using namespace std;

vector<mpz_class> eratosthenesSieve(int limit) {
    /* Строит решето Эратосфена */
    vector<bool> isPrime(limit + 1, true);
    isPrime[0] = isPrime[1] = false;

    for (int p = 2; p * p <= limit; p++) {
        if (isPrime[p]) {
            for (int i = p * p; i <= limit; i += p) {
                isPrime[i] = false;
            }
        }
    }

    vector<mpz_class> primes;
    for (int i = 2; i <= limit; ++i) {
        if (isPrime[i]) {
            primes.push_back(i);
        }
    }

    return primes;
}

vector<pair<mpz_class, mpz_class>> factorize(mpz_class num) {
    /* Находит каноническое разложение числа */
    vector<pair<mpz_class, mpz_class>> result;
    vector<mpz_class> primes = eratosthenesSieve(500);
    
    for (auto& primeElm : primes) {
        mpz_class exponent = 0;
        while (num % primeElm == 0 && num > 0) {
            exponent++;
            num /= primeElm;
        }
        if (exponent) {
            result.push_back({primeElm, exponent});
        }
    }

    return result;
}

bool millerTest(mpz_class n, int t) {
    /* Вероятностный тест Миллера */
    static gmp_randclass rand_gen(gmp_randinit_default);
    static random_device rd;
    rand_gen.seed(rd());

    mpz_class lower = 2;
    mpz_class upper = n - 1;

    mpz_class range = upper - lower + 1;

    if (n == 2 || n == 3 || n == 5 || n == 7) return true;
    if (n < 10) return false;

    set<mpz_class> nums;
    while ((int)nums.size() != t) {
        nums.insert(rand_gen.get_z_range(range) + lower);
    }

    for (auto& a : nums) {
        mpz_class base = a, exp = n - 1, mod = n, res = 0;
        mpz_powm(res.get_mpz_t(), base.get_mpz_t(), exp.get_mpz_t(), mod.get_mpz_t());
        if (res != 1) {
            return false;
        }
    }

    vector<pair<mpz_class, mpz_class>> primeFactors = factorize(n - 1);
    for (auto& factor : primeFactors) {
        mpz_class q = factor.first;
        bool onlyOne = true;
        for (auto& a : nums) {
            mpz_class base = a, exp = (n - 1) / q, mod = n, res = 0;
            mpz_powm(res.get_mpz_t(), base.get_mpz_t(), exp.get_mpz_t(), mod.get_mpz_t());
            if (res != 1) {
                onlyOne = false;
                break;
            }
        }
        if (onlyOne) {
            return false;
        }
    }
    return true;
}

mpz_class findGcd(mpz_class a, mpz_class b) {
    /* Ищет НОД(a, b) */
    if (b == 0) {
        return a;
    }

    return findGcd(b, a % b);
}

tuple<mpz_class, mpz_class, mpz_class> extendedEuclid(mpz_class a, mpz_class b) {
    /* Реализация расширенного алгоритма евклида */
    if (b == 0) {
        return {a, 1, 0};
    }

    mpz_class gcd = 0, u1 = 0, v1 = 0;
    tie(gcd, u1, v1) = extendedEuclid(b, a % b);

    mpz_class u = v1;
    mpz_class v = u1 - (a / b) * v1;

    return {gcd, u, v};
}

mpz_class modExp(mpz_class a, mpz_class x, mpz_class m) {
    /* Модульное возведение в степень */
    mpz_class res = 1;
    mpz_class base = a % m;
    while (x > 0) {
        if (x % 2 == 1) {
            res = (res * base) % m;
        }
        base = (base * base) % m;
        x /= 2;
    }
    return res;
}

mpz_class generateRandomPrime(int bitLength) {
    gmp_randclass rng(gmp_randinit_mt);
    rng.seed(time(nullptr));

    mpz_class candidate;
    while (true) {
        candidate = rng.get_z_bits(bitLength);
        mpz_setbit(candidate.get_mpz_t(), bitLength - 1);
        mpz_setbit(candidate.get_mpz_t(), 0);

        if (millerTest(candidate, 8)) {
            return candidate;
        }
    }
}

